R/LearnerAvg.R

```diff
@@ -163,35 +191,45 @@ LearnerRegrAvg = R6Class("LearnerRegrAvg", inherit = LearnerRegr,
 mlr_learners_regr.avg = NULL
 mlr_learners_classif.avg = NULL
 
-nlopt_levels = c("NLOPT_GN_DIRECT", "NLOPT_GN_DIRECT_L", "NLOPT_GN_DIRECT_L_RAND",
-  "NLOPT_GN_DIRECT_NOSCAL", "NLOPT_GN_DIRECT_L_NOSCAL", "NLOPT_GN_DIRECT_L_RAND_NOSCAL",
-  "NLOPT_GN_ORIG_DIRECT", "NLOPT_GN_ORIG_DIRECT_L", "NLOPT_LN_PRAXIS",
-  "NLOPT_GN_CRS2_LM", "NLOPT_GN_MLSL", "NLOPT_GN_MLSL_LDS",
-  "NLOPT_LN_COBYLA", "NLOPT_LN_NEWUOA", "NLOPT_LN_NEWUOA_BOUND",
-  "NLOPT_LN_NELDERMEAD", "NLOPT_LN_SBPLX", "NLOPT_LN_AUGLAG",
-  "NLOPT_LN_AUGLAG_EQ", "NLOPT_LN_BOBYQA", "NLOPT_GN_ISRES")
-
-nlopt_objfun = function(weights, task, measure, avg_weight_fun, data) {
-  # This is the objective function we minimize using nlopt
-  prd = avg_weight_fun(task, weights, data)
-  res = prd$score(measure)
-  if (measure$minimize) res else -res
-}
+optimize_weights_learneravg = function(self, task, n_weights, data) {
 
-optimize_objfun_nlopt = function(task, pars, avg_weight_fun, n_weights, data) {
-  require_namespaces("nloptr")
-  measure = assert_measure(as_measure(pars$measure, task_type = task$task_type))
+      # objective function to optimize
+      learneravg_objfun = function(x, task, measure, avg_weight_fun, data) {
+        # This is the objective function we minimize using nlopt
+        prd = avg_weight_fun(task, unlist(x), data)
+        res = prd$score(measure)
+        if (measure$minimize) res else -res
+      }
 
-  opt = nloptr::nloptr(
-    x0 = rep(1 / n_weights, n_weights),
-    eval_f = nlopt_objfun,
-    lb = rep(0, n_weights), ub = rep(1, n_weights),
-    eval_g_ineq = function(x, task, measure, avg_weight_fun, data) max(x) - 1,
-    opts = list(algorithm = pars$algorithm, xtol_rel = 1e-8),
-    task = task,
-    measure = measure,
-    avg_weight_fun = avg_weight_fun,
-    data = data
-  )$solution
+      pars = self$param_set$get_values(tags = "train")
+      ps = ParamSet$new(params = imap(data, function(x, n) {
+        if (is.numeric(n)) n = paste0("w.", n)
+        ParamDbl$new(id = n, lower = 0, upper = 1)
+      }))
+      optimizer = pars$optimizer
+      if (inherits(optimizer, "character")) {
+        optimizer = bbotk::opt(optimizer)
+        if (inherits(optimizer, "OptimizerNLoptr")) {
+          optimizer$param_set$values = list(xtol_rel = 1e-8, algorithm = "NLOPT_LN_COBYLA", x0 = rep(1/n_weights, n_weights))
```

Comment:
Maybe we want to set the x0 and xtol even if the optimizer was given as an argument (unless these values were set to something else by the user?) Or is that not possible? (i.e. is `x0` a "required" parameter and the user would always have to set these parameters?)

Reply:
`x0` is required for some algorithms e.g. COBYLA but not for all. 
We could do this, but I do not see much added benefit and we might run into strange behavior when e.g. a different algorithm is set.

I just wanted to have this be the default as it IIRC reflects what worked best in Ledell, 2015.